internal/mesh/internal/controllers/routes/generate.go

```diff
@@ -0,0 +1,631 @@
+
+		for _, port := range parentServiceDec.Data.Ports {
+			if port.Protocol == pbcatalog.Protocol_PROTOCOL_MESH {
+				inMesh = true
+				continue // skip
+			}
+			allowedPortProtocols[port.TargetPort] = port.Protocol
+		}
+	}
+
+	computedRoutes := &pbmesh.ComputedRoutes{
+		PortedConfigs: make(map[string]*pbmesh.ComputedPortRoutes),
+	}
+
+	// Visit all of the routes relevant to this computed routes.
+	routeNodesByPort := make(map[string][]*inputRouteNode)
+	related.WalkRoutesForParentRef(parentServiceRef, func(
+		rk resource.ReferenceKey,
+		res *pbresource.Resource,
+		xroute types.XRouteData,
+	) {
+		var (
+			ports          []string
+			wildcardedPort bool
+		)
+		for _, ref := range xroute.GetParentRefs() {
+			if resource.ReferenceOrIDMatch(ref.Ref, parentServiceRef) {
+				ports = append(ports, ref.Port)
+				if ref.Port == "" {
+					wildcardedPort = true
+					break
+				}
+			}
+		}
+
+		// Do a port explosion.
+		if wildcardedPort {
+			ports = nil
+			for port := range allowedPortProtocols {
+				ports = append(ports, port)
+			}
+		}
+
+		if len(ports) == 0 {
+			return // not relevant to this computed routes
+		}
+
+		for _, port := range ports {
+			if port == "" {
+				panic("impossible to have an empty port here")
+			}
+
+			// Check if the user provided port is actually valid.
+			nullRouteTraffic := (parentServiceDec == nil)
+			if _, ok := allowedPortProtocols[port]; !ok {
+				nullRouteTraffic = true
+			}
+
+			var node *inputRouteNode
+			switch route := xroute.(type) {
+			case *pbmesh.HTTPRoute:
+				if nullRouteTraffic {
+					node = newInputRouteNode(port)
+					setupDefaultHTTPRouteNode(node, types.NullRouteBackend)
+				} else {
+					node = compileHTTPRouteNode(port, res, route, related)
+				}
+			case *pbmesh.GRPCRoute:
+				if nullRouteTraffic {
+					node = newInputRouteNode(port)
+					setupDefaultGRPCRouteNode(node, types.NullRouteBackend)
+				} else {
+					node = compileGRPCRouteNode(port, res, route, related)
+				}
+			case *pbmesh.TCPRoute:
+				if nullRouteTraffic {
+					node = newInputRouteNode(port)
+					setupDefaultTCPRouteNode(node, types.NullRouteBackend)
+				} else {
+					node = compileTCPRouteNode(port, res, route, related)
+				}
+			default:
+				return // unknown xroute type (impossible)
+			}
+
+			routeNodesByPort[node.ParentPort] = append(routeNodesByPort[node.ParentPort], node)
+		}
+	})
+
+	// Fill in defaults where there was no xroute defined at all.
+	for port, protocol := range allowedPortProtocols {
+		if _, ok := routeNodesByPort[port]; !ok {
+			var typ *pbresource.Type
+			switch protocol {
+			case pbcatalog.Protocol_PROTOCOL_HTTP2:
+				typ = types.HTTPRouteType
+			case pbcatalog.Protocol_PROTOCOL_HTTP:
+				typ = types.HTTPRouteType
+			case pbcatalog.Protocol_PROTOCOL_GRPC:
+				typ = types.GRPCRouteType
+			case pbcatalog.Protocol_PROTOCOL_TCP:
+				typ = types.TCPRouteType
+			default:
+				typ = types.TCPRouteType
+			}
```

Comment:
I think this case is not tested. There are also some lines below that are untested based on my test coverage run.